File: shell_wrapper/shell_types.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "shell_wrapper/shell_types.h"

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <memory>
#include <utility>
#include <vector>

#include "absl/log/log.h"
#include "absl/numeric/int128.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/types/span.h"
#include "shell_encryption/int256.h"
#include "shell_encryption/rns/crt_interpolation.h"
#include "shell_wrapper/shell_aliases.h"
#include "shell_wrapper/shell_types.rs.h"
#include "shell_wrapper/status.h"
#include "shell_wrapper/status.rs.h"

namespace secure_aggregation {

RnsContextConfig ParseRnsContextConfig(uint64_t log_n, uint64_t t,
                                       const uint64_t* qs, uint64_t num_qs) {
  RnsContextConfig config = {
      .log_n = static_cast<int>(log_n), .qs = {}, .t = t};
  if (qs == nullptr) {
    return config;
  }
  for (int i = 0; i < num_qs; ++i) {
    config.qs.push_back(qs[i]);
  }
  return config;
}

}  // namespace secure_aggregation

using secure_aggregation::MakeFfiStatus;

FfiStatus WriteSmallRnsPolynomialToBuffer(const RnsPolynomialWrapper* poly,
                                          ModuliWrapper moduli,
                                          uint64_t buffer_len, int64_t* buffer,
                                          uint64_t* n_written) {
  if (poly == nullptr || poly->ptr == nullptr || buffer == nullptr ||
      n_written == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  if (poly->ptr->IsNttForm()) {
    absl::Status status =
        poly->ptr->ConvertToCoeffForm({moduli.moduli, moduli.len});
    if (!status.ok()) {
      return MakeFfiStatus(status);
    }
  }

  for (int i = 0; i < moduli.len; ++i) {
    const auto& coeffs = poly->ptr->Coeffs()[i];
    uint64_t q = moduli.moduli[i]->Modulus();
    uint64_t q_half = q >> 1;

    if (i == 0) {
      // We write to the buffer only for the first modulus.
      *n_written = std::min(static_cast<size_t>(buffer_len), coeffs.size());
    }

    for (int j = 0; j < *n_written; ++j) {
      // For a small integer c, |c| mod q = |c|.
      uint64_t value = coeffs[j].ExportInt(moduli.moduli[i]->ModParams());
      int64_t signed_value;
      if (value < q_half) {
        signed_value = static_cast<int64_t>(value);
      } else {
        // Larger values (but still smaller than q, since ExportInt outputs in
        // the range [0, q]) are represented as negative integers. E.g. q =
        // 1125899906826241, value = 1125899906826240 gives signed_value = -1.
        signed_value = -static_cast<int64_t>(q - value);
      }

      if (i == 0) {
        // Write down to buffer. Technically we only need the first modulus.
        buffer[j] = signed_value;
      } else {
        // But check that the other moduli are consistent with the first one.
        if (signed_value != buffer[j]) {
          return MakeFfiStatus(absl::InvalidArgumentError(
              "Coefficients don't match across moduli, not a small "
              "polynomial."));
        }
      }
    }
  }

  return MakeFfiStatus();
}

FfiStatus ReadSmallRnsPolynomialFromBuffer(const int64_t* buffer,
                                           uint64_t buffer_len, uint64_t log_n,
                                           ModuliWrapper moduli,
                                           RnsPolynomialWrapper* out) {
  if (buffer == nullptr || out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  int num_coeffs = 1 << log_n;
  if (buffer_len > num_coeffs) {
    return MakeFfiStatus(
        absl::InvalidArgumentError("Buffer has too many coefficients, it does "
                                   "not fit in one polynomial."));
  }

  int num_moduli = moduli.len;
  std::vector<std::vector<secure_aggregation::ModularInt>> coeffs(num_moduli);
  for (int j = 0; j < num_coeffs; ++j) {
    int64_t value = 0;  // Pad unused coeffs with 0s.
    if (j < buffer_len) {
      value = buffer[j];
    }
    // Convert `value` to balanced representation mod q_i for each q_i.
    for (int i = 0; i < num_moduli; ++i) {
      auto q_i = moduli.moduli[i];
      uint64_t unsigned_value;
      if (value >= 0) {
        unsigned_value = static_cast<uint64_t>(value);
      } else {
        // -q_i < value < 0 for small polynomial, so r.h.s. is positive.
        unsigned_value =
            static_cast<uint64_t>(static_cast<int64_t>(q_i->Modulus()) + value);
      }
      auto value_mod_qi = secure_aggregation::ModularInt::ImportInt(
          unsigned_value, q_i->ModParams());
      if (!value_mod_qi.ok()) {
        return MakeFfiStatus(value_mod_qi.status());
      }
      coeffs[i].push_back(value_mod_qi.value());
    }
  }
  auto poly = secure_aggregation::RnsPolynomial::Create(std::move(coeffs),
                                                        /*is_ntt=*/false);
  if (!poly.ok()) {
    return MakeFfiStatus(poly.status());
  }
  out->ptr = std::make_unique<secure_aggregation::RnsPolynomial>(
      std::move(poly.value()));
  return MakeFfiStatus();
}

FfiStatus AddInPlace(ModuliWrapper moduli, const RnsPolynomialWrapper* in,
                     RnsPolynomialWrapper* out) {
  if (in == nullptr || out == nullptr || in->ptr == nullptr ||
      out->ptr == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }
  if (absl::Status status =
          out->ptr->AddInPlace(*in->ptr, {moduli.moduli, moduli.len});
      !status.ok()) {
    return MakeFfiStatus(std::move(status));
  }
  return MakeFfiStatus();
}

FfiStatus AddInPlaceVec(ModuliWrapper moduli, const RnsPolynomialVecWrapper* in,
                        RnsPolynomialVecWrapper* out) {
  if (in == nullptr || out == nullptr || in->ptr == nullptr ||
      out->ptr == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }
  if (in->ptr->size() != out->ptr->size()) {
    return MakeFfiStatus(
        absl::InvalidArgumentError("`in` and `out` must have the same size."));
  }
  for (int i = 0; i < in->ptr->size(); ++i) {
    absl::Status status =
        out->ptr->at(i).AddInPlace(in->ptr->at(i), {moduli.moduli, moduli.len});
    if (!status.ok()) {
      return MakeFfiStatus(std::move(status));
    }
  }
  return MakeFfiStatus();
}

FfiStatus WriteRnsPolynomialToBuffer128(
    const secure_aggregation::RnsContext* rns_context,
    const RnsPolynomialWrapper* poly, uint64_t buffer_len, uint64_t* buffer) {
  if (buffer == nullptr || poly == nullptr || rns_context == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }
  const uint64_t expected_buffer_len = 2 * poly->ptr->NumCoeffs();
  if (buffer_len != expected_buffer_len) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        absl::StrCat("Expected `buffer_len` == ", expected_buffer_len,
                     " (poly->ptr->NumCoeffs() * 2), but got ", buffer_len)));
  }

  // Convert poly to coefficient form if necessary.
  RnsPolynomialWrapper poly_coefficient_form;
  if (poly->ptr->IsNttForm()) {
    poly_coefficient_form.ptr =
        std::make_unique<secure_aggregation::RnsPolynomial>(*poly->ptr);
    absl::Status status = poly_coefficient_form.ptr->ConvertToCoeffForm(
        rns_context->MainPrimeModuli());
    if (!status.ok()) {
      return MakeFfiStatus(status);
    }
    poly = &poly_coefficient_form;
  }

  // Compute `modulus_hats` and `modulus_hat_invs`.
  auto moduli = rns_context->MainPrimeModuli();
  absl::StatusOr<std::vector<rlwe::uint256>> modulus_hats =
      rlwe::RnsModulusComplements<secure_aggregation::ModularInt,
                                  rlwe::uint256>(moduli);
  if (!modulus_hats.ok()) {
    return MakeFfiStatus(modulus_hats.status());
  }
  absl::StatusOr<std::vector<secure_aggregation::ModularInt>> modulus_hat_invs =
      rns_context->MainPrimeModulusCrtFactors(moduli.size() - 1);
  if (!modulus_hat_invs.ok()) {
    return MakeFfiStatus(modulus_hat_invs.status());
  }

  // Interpolate `poly`.
  absl::StatusOr<std::vector<rlwe::uint256>> interpolated_coeffs =
      rlwe::CrtInterpolation<secure_aggregation::ModularInt, rlwe::uint256>(
          poly->ptr->Coeffs(), moduli, *modulus_hats, *modulus_hat_invs);
  if (!interpolated_coeffs.ok()) {
    return MakeFfiStatus(interpolated_coeffs.status());
  }

  // Safety check that the number of interpolated coefficients doesn't overflow
  // the buffer. Should always pass as long as CrtInterpolation is correct,
  // since `buffer_len` is validated above.
  if (interpolated_coeffs->size() * 2 > buffer_len) {
    return MakeFfiStatus(absl::InternalError(absl::StrCat(
        "2 * interpolated_coeffs->size() (=", interpolated_coeffs->size(),
        ") > buffer_len (=", buffer_len, ")")));
  }
  // Write coefficients to buffer.
  for (int i = 0; i < interpolated_coeffs->size(); ++i) {
    const auto& coeff = (*interpolated_coeffs)[i];
    if (rlwe::Uint256High128(coeff) != 0) {
      return MakeFfiStatus(absl::InvalidArgumentError(
          "Coefficient cannot be larger than 128 bits"));
    }
    absl::uint128 coeff_128 = rlwe::Uint256Low128(coeff);
    buffer[2 * i] = absl::Uint128Low64(coeff_128);
    buffer[2 * i + 1] = absl::Uint128High64(coeff_128);
  }

  return {};
}

RnsPolynomialWrapper CloneRnsPolynomialWrapper(const RnsPolynomialWrapper* in) {
  return RnsPolynomialWrapper{
      .ptr = std::make_unique<secure_aggregation::RnsPolynomial>(
          in->ptr->Clone())};
}

RnsPolynomialVecWrapper CloneRnsPolynomialVecWrapper(
    const RnsPolynomialVecWrapper* in) {
  RnsPolynomialVecWrapper result{
      .len = in->len,
      .ptr = std::make_unique<std::vector<secure_aggregation::RnsPolynomial>>(),
  };
  result.ptr->reserve(in->ptr->size());
  for (const auto& poly : *in->ptr) {
    result.ptr->push_back(poly.Clone());
  }
  return result;
}

File: shell_wrapper/shell_types.rs
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use status::Status;
use std::marker::PhantomData;
use std::mem::MaybeUninit;

#[cxx::bridge]
pub mod ffi {
    /// Owned RnsPolynomial behind a unique_ptr.
    pub struct RnsPolynomialWrapper {
        pub ptr: UniquePtr<RnsPolynomial>,
    }

    pub struct RnsPolynomialVecWrapper {
        /// Number of polynomials in the vector, for easy access from Rust.
        len: i32,
        ptr: UniquePtr<CxxVector<RnsPolynomial>>,
    }

    /// Stores a span of pointers to the prime moduli, for polynomial operations. The
    /// pointers are backed by a AhePublicParameters or KahePublicParameters struct.
    #[derive(Copy, Clone)]
    struct ModuliWrapper {
        moduli: *const *const PrimeModulus,
        len: usize,
    }

    unsafe extern "C++" {
        type FfiStatus = status::ffi::FfiStatus;
        include!("shell_wrapper/shell_types.h");
        include!("shell_wrapper/shell_aliases.h");
        #[namespace = "secure_aggregation"]
        type RnsPolynomial;
        #[namespace = "secure_aggregation"]
        type RnsContext;

        #[namespace = "secure_aggregation"]
        type PrimeModulus;

        pub fn CreateEmptyRnsPolynomialWrapper() -> RnsPolynomialWrapper;

        pub unsafe fn WriteSmallRnsPolynomialToBuffer(
            poly: *const RnsPolynomialWrapper,
            moduli: ModuliWrapper,
            buffer_len: u64,
            buffer: *mut i64,
            n_written: *mut u64,
        ) -> FfiStatus;

        pub unsafe fn ReadSmallRnsPolynomialFromBuffer(
            buffer: *const i64,
            buffer_len: u64,
            log_n: u64,
            moduli: ModuliWrapper,
            out: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;

        pub unsafe fn AddInPlace(
            moduli: ModuliWrapper,
            in_: *const RnsPolynomialWrapper,
            out: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;

        pub unsafe fn AddInPlaceVec(
            moduli: ModuliWrapper,
            in_: *const RnsPolynomialVecWrapper,
            out: *mut RnsPolynomialVecWrapper,
        ) -> FfiStatus;

        pub unsafe fn WriteRnsPolynomialToBuffer128(
            rns_context: *const RnsContext,
            poly: *const RnsPolynomialWrapper,
            buffer_len: usize,
            buffer: *mut u64,
        ) -> FfiStatus;

        pub unsafe fn CloneRnsPolynomialWrapper(
            poly: *const RnsPolynomialWrapper,
        ) -> RnsPolynomialWrapper;

        pub unsafe fn CloneRnsPolynomialVecWrapper(
            poly: *const RnsPolynomialVecWrapper,
        ) -> RnsPolynomialVecWrapper;

        pub fn CloneString(x: &CxxString) -> UniquePtr<CxxString>;
        pub fn EmptyString() -> &'static CxxString;

    }
}

pub use ffi::ModuliWrapper;
use status::rust_status_from_cpp;

/// Re-export CXX bindings, and implement Clone for them.
pub use ffi::RnsPolynomialVecWrapper as RnsPolynomialVec;

impl Clone for RnsPolynomialVec {
    fn clone(&self) -> Self {
        // SAFETY: No lifetime constraints. self->ptr is guaranteed to be non-null.
        unsafe { ffi::CloneRnsPolynomialVecWrapper(self) }
    }
}

pub use ffi::RnsPolynomialWrapper as RnsPolynomial;

impl Clone for RnsPolynomial {
    fn clone(&self) -> Self {
        // SAFETY: No lifetime constraints. self->ptr is guaranteed to be non-null.
        unsafe { ffi::CloneRnsPolynomialWrapper(self) }
    }
}

pub use ffi::CreateEmptyRnsPolynomialWrapper as create_empty_rns_polynomial;

/// Store some raw pointers to the prime moduli. The bindings don't know that
/// ModuliWrapper actually has a lifetime, so we add it manually with
/// PhantomData.
pub struct Moduli<'a> {
    pub moduli: ModuliWrapper, // Only valid for the lifetime of the Ahe or Kahe parameters.
    pub phantom: PhantomData<&'a ()>,
}

/// Represents a borrowed reference to an RnsContext. Used for polynomial
/// operations. The bindings don't know that RnsContext actually has a
/// lifetime, so we add it manually with PhantomData.
pub struct RnsContextRef<'a> {
    pub rns_context: *const ffi::RnsContext, /* Only valid for the lifetime of the Ahe or Kahe
                                              * parameters. */
    pub phantom: PhantomData<&'a ()>,
}

///  Converts slices into C++-safe pointer and length pairs.
pub fn to_cpp_pointer_len_pair<T>(s: &[T]) -> (*const T, usize) {
    if s.is_empty() {
        (std::ptr::null(), 0)
    } else {
        (s.as_ptr(), s.len())
    }
}

/// Converts slices into C++-safe mutable pointer and length pairs.
pub fn to_cpp_pointer_len_pair_mut<T>(s: &mut [T]) -> (*mut T, usize) {
    if s.is_empty() {
        (std::ptr::null_mut(), 0)
    } else {
        (s.as_mut_ptr(), s.len())
    }
}

/// Takes prime moduli {q_i}, and the RNS representation `poly` of a "small"
/// polynomial in Z[X] where each coefficient c \in Z verifies |c| < q_i/2 for
/// all q_i. Fills in the buffer with the signed integer coefficients of `poly`.
/// Accepts `poly` in both coefficient form and NTT form (by internally
/// converting to coefficient form if needed). Returns the number of buffer
/// elements written if successful, and a StatusError otherwise.
pub fn write_small_rns_polynomial_to_buffer(
    poly: &RnsPolynomial,
    moduli: &Moduli,
    buffer: &mut [i64],
) -> Result<usize, status::StatusError> {
    let mut n_written: u64 = 0;
    // SAFETY: No lifetime constraints. Writes to `buffer` within a valid range.
    rust_status_from_cpp(unsafe {
        ffi::WriteSmallRnsPolynomialToBuffer(
            poly,
            moduli.moduli,
            buffer.len() as u64,
            buffer.as_mut_ptr(),
            &mut n_written,
        )
    })?;
    Ok(n_written as usize)
}

/// Takes prime moduli {q_i}, and a buffer of `buffer_len` signed integers,
/// representing the coefficients of a "small" polynomial in Z[X] where each
/// coefficient c \in Z verifies |c| < q_i/2 for all q_i. Returns a
/// RnsPolynomialWrapper containing the polynomial in RNS coefficient form if
/// successful, and a StatusError otherwise.
pub fn read_small_rns_polynomial_from_buffer(
    buffer: &[i64],
    log_n: u64,
    moduli: &Moduli,
) -> Result<RnsPolynomial, status::StatusError> {
    let mut out = MaybeUninit::<RnsPolynomial>::zeroed();
    // SAFETY: No lifetime constraints (the resulting `RnsPolynomial` does not keep
    // any references). Reads from `buffer` within a valid range. `out` is safely
    // initialized if `status` is OK.
    rust_status_from_cpp(unsafe {
        ffi::ReadSmallRnsPolynomialFromBuffer(
            buffer.as_ptr(),
            buffer.len() as u64,
            log_n,
            moduli.moduli,
            out.as_mut_ptr(),
        )
    })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

/// Adds the polynomial `in` to `out` in-place, using the RNS `moduli`. Returns
/// () if successful, and a StatusError otherwise.
pub fn add_in_place(
    moduli: &Moduli,
    in_polynomial: &RnsPolynomial,
    out_polynomial: &mut RnsPolynomial,
) -> Status {
    // SAFETY: No lifetime constraints (`AddInPlace` copies the contents of
    // `in_polynomial` and adds them into `out_polynomial`).
    rust_status_from_cpp(unsafe { ffi::AddInPlace(moduli.moduli, in_polynomial, out_polynomial) })
}

// Adds the vector of polynomials `in` to `out` in-place element-wise, using the
// RNS `moduli`. Does not failatomically: if the result is not Ok, the value of
// `out` is undefined.
pub fn add_in_place_vec(
    moduli: &Moduli,
    in_polynomial_vec: &RnsPolynomialVec,
    out_polynomial_vec: &mut RnsPolynomialVec,
) -> Status {
    // SAFETY: No lifetime constraints (`AddInPlaceVec` copies the contents of
    // `in_polynomial_vec` and adds them into `out_polynomial_vec`).
    rust_status_from_cpp(unsafe {
        ffi::AddInPlaceVec(moduli.moduli, in_polynomial_vec, out_polynomial_vec)
    })
}

/// Converts the given RnsPolynomial `poly` to coefficient form, interpolating
/// the coefficients to a single modulus using CRT interpolation. Writes the
/// resulting coefficient vector to `buffer`, using two consecutive uint64_t
/// words for every coefficient, with the lower half being written first.
/// Returns an error if any pointer arguments are null, if any coefficient
/// exceeds 128 bits, or if the buffer length is not equal to
/// `2*poly->ptr->NumCoeffs()`.
pub fn write_rns_polynomial_to_buffer_128(
    rns_context: &RnsContextRef,
    in_polynomial: &RnsPolynomial,
    output_buffer: &mut [u64],
) -> status::Status {
    let (out_ptr, out_len) = to_cpp_pointer_len_pair_mut(output_buffer);
    // SAFETY: No lifetime constraints. The number of elements written is checked
    // inside the C++ function.
    rust_status_from_cpp(unsafe {
        ffi::WriteRnsPolynomialToBuffer128(rns_context.rns_context, in_polynomial, out_len, out_ptr)
    })?;
    Ok(())
}
